src/main/scala/viper/gobra/translator/encodings/channels/ChannelEncoding.scala

```diff
@@ -0,0 +1,248 @@
+
+  /**
+    * Translates a type into a Viper type.
+    */
+  override def typ(ctx : Context) : in.Type ==> vpr.Type = {
+    case ctx.Channel(t) / m =>  m match {
+      case Exclusive => vpr.Int // for now we simply map it to Viper Ints
+      case Shared => vpr.Ref
+    }
+  }
+
+  /**
+    * Encodes expressions as values that do not occupy some identifiable location in memory.
+    *
+    * R[ dflt(channel[T]) ] -> 0
+    * R[ nil : channel[T] ] -> 0
+    * R[ <- (c: channel[T] ] ->
+    *   exhale acc([c].RecvChannel(), wildcard)
+    *   exhale [c].RecvGivenPerm()()
+    *   var res [ T ]
+    *   inhale [c].RecvChannel()
+    *   res
+    */
+  override def expr(ctx : Context) : in.Expr ==> CodeWriter[vpr.Exp] = {
+    default(super.expr(ctx)) {
+      case (exp : in.DfltVal) :: ctx.Channel(t) / Exclusive =>
+        unit(withSrc(vpr.IntLit(0), exp))
+
+      case (lit: in.NilLit) :: ctx.Channel(t) =>
+        unit(withSrc(vpr.IntLit(0), lit))
+
+      case exp@in.Receive(channel :: ctx.Channel(typeParam), recvChannel, recvGivenPerm) =>
+        val (pos, info, errT) = exp.vprMeta
+        val res = in.LocalVar(Names.freshName, typeParam.withAddressability(Addressability.Exclusive))(exp.info)
+        val vprRes = ctx.typeEncoding.variable(ctx)(res)
+        val recvChannelPred = in.Accessible.Predicate(in.MPredicateAccess(channel, recvChannel, Vector())(exp.info))
+        val recvChannelWildcard = in.Access(recvChannelPred, in.WildcardPerm(exp.info))(exp.info)
+        for {
+          // exhale acc([c].RecvChannel(), wildcard)
+          vprRecvChannelWildcard <- ctx.ass.translate(recvChannelWildcard)(ctx)
+          vprExhaleRecvChannelWildcard = vpr.Exhale(vprRecvChannelWildcard)(pos, info, errT)
+          _ <- write(vprExhaleRecvChannelWildcard)
+
+          // exhale [c].RecvGivenPerm()()
+          recvGivenPermInst = getChannelInvariantAccess(channel, recvGivenPerm, Vector())(exp.info)
+          vprRecvGivenPermInst <- ctx.ass.translate(recvGivenPermInst)(ctx)
+          vprExhaleRecvGivenPermInst = vpr.Exhale(vprRecvGivenPermInst)(pos, info, errT)
+          _ <- write(vprExhaleRecvGivenPermInst)
+
+          // var res [ T ]
+          _ <- local(vprRes)
+
+          // inhale [c].RecvChannel()
+          recvChannelFull = in.Access(recvChannelPred, in.FullPerm(exp.info))(exp.info)
+          vprRecvChannelFull <- ctx.ass.translate(recvChannelFull)(ctx)
+          vprInhaleRecvChannelFull = vpr.Inhale(vprRecvChannelFull)(pos, info, errT)
+          _ <- write(vprInhaleRecvChannelFull)
+
+          // res
+          vprResRead <- ctx.typeEncoding.expr(ctx)(res)
+        } yield vprResRead
+    }
+  }
+
+  /**
+    * Encodes
+    *  - allocation of a new channel
+    *  - channel receive operation that returns received message as well as a success flag
+    *
+    * [r := make(chan T, bufferSize)] ->
+    *   assert 0 <= [bufferSize]
+    *   var a [ chan T ]
+    *   inhale [a].isChannel([bufferSize])
+    *   r := a
```

Comment:
In the current encoding, we cannot derive that a channel is not nil if we have some permissions associated to it. Shortterm this might not matter, but longterm we should figure this out. Maybe, if you are not fixing this now, could you please create an issue.

Reply:
As we encoding the channel type as an int would this mean that we ‚reserve‘ 0 as the channel‘s nil value and would add contracts that ensure that the channel is != 0?

Reply:
I am not sure if I would call it "adding contracts". For Ref, we know that `acc(x.f)` implies that `x` is not null. This could also be achieved by replacing `acc(x.f)` with `acc(x.f) && x != null`. In theory, we could try the same here. But there are different approaches as well.

Reply:
Issue #174 created